grafana/grafana
# Explore: Map Graphite queries to Loki

**What this PR does / why we need it**:

This will allow to map specific metric nodes from a Graphite query to Loki labels when data source is switched. It's a simplified version of syntax suggested in #21205 but can be extended in future if it's needed.

**Which issue(s) this PR fixes**:



Fixes #21205

**Special notes for your reviewer**:

The main logic for mapping is in: `public/app/plugins/datasource/loki/importing/fromGraphite.ts`. It's tested with an integration test `public/app/plugins/datasource/loki/importing/importing.test.ts` covering parsing settings, parsing queries and executing the conversion.

*How to test it?*

1. Run: `make devenv sources=graphite,loki`

2. Create a new Graphite data source pointing to `http://localhost:8180` (provisioned cannot be edited)

3. Add a mapping: `servers.(cluster).(server).*` in data source configuration:

<details><summary>show screenshot</summary>
<img width="967" alt="Screenshot 2021-04-27 at 09 52 00" src="https://user-images.githubusercontent.com/745532/116205669-41484b80-a73e-11eb-89bf-353d892fe72e.png">
</details>

4. Save the data source.

6. Run [graphite-and-loki.sh](https://gist.github.com/ifrost/2a18248d588f7d46e82245085f189315) script to generate metrics and logs for testing.

7. Open newly created data source in Explore, switch to last 15 minutes view and run the following query `servers.*.*.cpu` to see if the data was ingested properly. You should see something like this:

<details><summary>show screenshot</summary>
<img width="1462" alt="Screenshot 2021-04-27 at 09 33 39" src="https://user-images.githubusercontent.com/745532/116203899-78b5f880-a73c-11eb-8fa6-469146bbb0c8.png">
</details>

8. Modify the query `servers.west.002.*` and switch to Loki - the query should be mapped:

<details><summary>show screenshot</summary>

![map](https://user-images.githubusercontent.com/745532/116204512-22958500-a73d-11eb-9365-9d5163bf608a.gif)

</details>

Looping in: @Dieterbe
